contracts/LimitsChecker.sol

```diff
@@ -0,0 +1,267 @@
+// SPDX-FileCopyrightText: 2022 Lido <info@lido.fi>
+// SPDX-License-Identifier: GPL-3.0
+
+pragma solidity ^0.8.4;
+
+import "./libraries/EVMScriptCreator.sol";
+import "./interfaces/IBokkyPooBahsDateTimeContract.sol";
+import "OpenZeppelin/openzeppelin-contracts@4.3.2/contracts/access/AccessControl.sol";
+
+/// @author zuzueeka
+/// @notice Stores limits params and checks limits
+///
+/// ▲ limit-spent
+/// |
+/// │......               |.............       |..............     limit-spent = limit
+/// │      .......        |                    |
+/// │             ....    |             ....   |
+/// │                 ....|                 ...|
+/// │─────────────────────────────────────────────────────────────> Time
+/// |      ^      ^   ^   |             ^   ^  |                 ^ - Motion enactment
+/// │                     |currentPeriodEnd    |currentPeriodEnd
+/// |                     |spent=0             |spent=0
+///
+/// currentPeriodEnd is calculated as a calendar date of the beginning of a next month, bi-months, quarter, half year, or year period.
+/// If, for example, periodDurationMonth = 3, then it is considered that the date changes once a quarter. And can take values 01.01, 01.04, 01.07, 01.10.
+/// If periodDurationMonth = 1, then shift of currentPeriodEnd occures once a month and currentPeriodEnd can take values 01.01, 01.02, 01.03 etc
+///
+contract LimitsChecker is AccessControl {
+    // -------------
+    // EVENTS
+    // -------------
+    event LimitsParametersChanged(uint256 _limit, uint256 _periodDurationMonth);
+    event FundsSpent(
+        uint256 _alreadySpentAmount,
+        uint256 _spendableAmount,
+        uint256 _periodStartTimestamp,
+        uint256 _periodEndTimestamp
+    );
+
+    // -------------
+    // ERRORS
+    // -------------
+    string private constant ERROR_WRONG_PERIOD_DURATION = "WRONG_PERIOD_DURATION";
+    string private constant ERROR_SUM_EXCEEDS_LIMIT = "SUM_EXCEEDS_LIMIT";
+    // -------------
+    // ROLES
+    // -------------
+    bytes32 public constant SET_LIMIT_PARAMETERS_ROLE = keccak256("SET_LIMIT_PARAMETERS_ROLE");
+    bytes32 public constant UPDATE_LIMIT_SPENDINGS_ROLE = keccak256("UPDATE_LIMIT_SPENDINGS_ROLE");
+
+    // -------------
+    // CONSTANTS
+    // -------------
+
+    // ------------
+    // STORAGE VARIABLES
+    // ------------
+
+    /// @notice Address of BokkyPooBahsDateTimeContract
+    IBokkyPooBahsDateTimeContract public immutable bokkyPooBahsDateTimeContract;
+
+    /// @notice Length of period in months
+    uint64 internal periodDurationMonth;
+
+    /// @notice End of the current period
+    uint128 internal currentPeriodEnd;
+
+    /// @notice The maximum that can be spent in a period
+    uint128 internal limit;
+
+    /// @notice Amount already spent in the period
+    uint128 internal spent;
```

Comment:
maybe `spent` --> `spending`
a bit more correct as far as I understand
and will correspond to "spending" in `UPDATE_LIMIT_SPENDINGS_ROLE` etc
but then need to update `_alreadySpentAmount` (e. g. to `_currentSpending`) etc

Reply:
renamed:
spent -> spentAmount
UPDATE_LIMIT_SPENDINGS_ROLE -> UPDATE_SPENT_AMOUNT_ROLE
_spendableAmount -> _spendableBalance